### PR TITLE
Client: Change spider to access unvisited URLs

### PR DESCRIPTION
**Pull request by psiinon**
## Overview
In ZAP with the client add-on installed open the browser and point at a running version of bodgeit.
Then run the client spider against bodgeit - it will not find anything.
This is because its listenning for "new node" events, but no new nodes are added because they were already there, just unvisited.
This PR fixes that by adding those nodes to the task list.

No changelog entry as this is still unpublished functionality 😁 

## Checklist
- [ ] Update help
- [ ] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [x] Check code coverage
- [x] Sign-off commits
- [ ] Squash commits
- [x] Use a descriptive title

### CODE REVIEW

File: addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java [hunk 1]
```diff
@@ -318,6 +309,16 @@ public void eventReceived(Event event) {
                         eventConsumer, "org.zaproxy.zap.extension.spiderAjax.SpiderEventPublisher");
     }
 
+    private ClientMap getClientTree() {
+        clientTree =
+                new ClientMap(
+                        new ClientNode(
+                                new ClientSideDetails(
+                                        Constant.messages.getString("client.tree.title"), null),
+                                this.getModel().getSession()));
```
Comment (kingthorin) [inline addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java:318]: Do we really want this instantiating new objects every time it's called?

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpider.java [hunk 2]
```diff
@@ -111,7 +116,36 @@ public void start() {
                         new ClientSpiderThreadFactory(
                                 "ZAP-ClientSpiderThreadPool-" + id + "-thread-"));
 
+        List<String> unvisitedUrls = getUnvisitedUrls();
+
         addTask(targetUrl, options.getInitialLoadTimeInSecs());
+
+        // Add all of the known but unvisited URLs otherwise these will get ignored
+        unvisitedUrls.forEach(url -> addTask(url, options.getInitialLoadTimeInSecs()));
```
Comment (kingthorin) [inline addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpider.java:124]: Shouldn't this be a collection that's shared across spiders/threads. Won't they only be "unvisited" for the current browser, though not necessarily for _n_ running spiders?

File: addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java (hunk 1 above)
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/ExtensionClientIntegration.java]: Yeah, thats a bug - fixed :)

File: addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpider.java (hunk 2 above)
Comment (psiinon) [reply addOns/client/src/main/java/org/zaproxy/addon/client/spider/ClientSpider.java:124]: They are only unvisited at the moment that the spider is started. they could be visited at any point after that, so I dont think anything needs to be shared.

**Comment by psiinon** [comment]
Interesting - the unit tests pass in Eclipse.
But they fail in the same way as here from the command line, so will look into that..

**Comment by psiinon** [comment]
Tests fixed

**Review by kingthorin — APPROVED** [review]

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**